File: src/main/java/fluid/physics/Universe.java
package fluid.physics;

import fluid.entity.FluidEntity;

import java.util.stream.IntStream;

/**
 * http://cowboyprogramming.com/2008/04/01/practical-fluid-mechanics/
 * http://www.dgp.toronto.edu/people/stam/reality/Research/pdf/GDC03.pdf
 */
public class Universe {

    /**
     * Maximum number of immediate neighbors that each {@link FluidEntity} may have bidirectional interactions with.
     * Used to ensure that an entity does not transfer more than 100% of its heat/pressure/etc to its neighbors
     */
    public static final int MAX_NEIGHBORS = 4;

    private int step = 0;

    private final FluidEntity[][] entities;

    public Universe(FluidEntity[][] entities) {
        this.entities = entities;
    }

    /**
     * Run round of physics
     */
    public synchronized void updateUniverseState() {
        ExternalInput.applyInput(entities, step);
        incrementFluid();
        step++;
    }

    public FluidEntity[][] getEntities() {
        return entities;
    }

    private static final double GRAVITATIONAL_CONSTANT = .00001;

    private void incrementFluid() {
        // force applications
        applyBidirectionalInteractions();
        applyGravity();

        IntStream.range(0, entities.length).forEach(x -> IntStream.range(0, entities[x].length).forEach(y -> entities[x][y].changeHeat()));
        IntStream.range(0, entities.length).forEach(x -> IntStream.range(0, entities[x].length).forEach(y -> entities[x][y].changeForce()));

        // transfer logging
        advection();

        // transfer application
        IntStream.range(0, entities.length).forEach(x -> IntStream.range(0, entities[x].length).forEach(y -> entities[x][y].convertMassTransferToAbsoluteChange()));
        IntStream.range(0, entities.length).parallel().forEach(x -> IntStream.range(0, entities[x].length).forEach(y -> entities[x][y].changeMass()));
    }


    private void applyBidirectionalInteractions() {
        IntStream.range(0, entities.length - 1).parallel().forEach(i -> IntStream.range(0, entities[i].length - 1).forEach(j -> {
            FluidEntity entity = entities[i][j];

            // Right entity
            entity.applyBidirectionalInteractions(entities[i + 1][j]);

            // Upper entity
            entity.applyBidirectionalInteractions(entities[i][j + 1]);
        }));
    }

    private void applyGravity() {
        final double[][] downwardPressure = new double[entities.length][entities[0].length];

        for (int i = 0; i < entities.length; i++) {
            for (int j = entities[i].length - 2; j >= 0; j--) {
                downwardPressure[i][j] = downwardPressure[i][j + 1] + (entities[i][j].getMass() * GRAVITATIONAL_CONSTANT);
            }
        }

        IntStream.range(0, entities.length).parallel().forEach(x -> IntStream.range(0, entities[x].length).forEach(y -> entities[x][y].addForceY(-downwardPressure[x][y])));
    }
    
    /**
     * Advection moves the quantities from point to its connections/neighbors. Quantities include velocity/mass/heat/etc.
     * The amount moved from one point to another is based on the given point's velocity.
     */
    private void advection() {
        IntStream.range(0, entities.length).parallel().forEach(i -> IntStream.range(0, entities[i].length).forEach(j -> {
            forwardAdvectionCellTransfer(i, j);
            reverseAdvectionCellTransfer(i, j);
        }));
    }

    /**
     * https://en.wikipedia.org/wiki/Bilinear_interpolation
     */
    private void forwardAdvectionCellTransfer(int xIndex, int yIndex) {
        FluidEntity entity = entities[xIndex][yIndex];
        double deltaX = entity.getDeltaX();
        double deltaY = entity.getDeltaY();

        if (deltaX == 0 && deltaY == 0) return;

        int xIndexOffset = (int) deltaX / FluidEntity.SPACE;
        int yIndexOffset = (int) deltaY / FluidEntity.SPACE;

        boolean dxPositive = deltaX > 0;
        boolean dyPositive = deltaY > 0;

        int t1x = getLesserTargetIndex(xIndex, xIndexOffset, dxPositive);
        int t1y = getLesserTargetIndex(yIndex, yIndexOffset, dyPositive);

        int t2x = t1x + 1;
        int t2y = t1y + 1;

        double xPosInCell = dxPositive ? entity.getDeltaX() % FluidEntity.SPACE : FluidEntity.SPACE + entity.getDeltaX() % FluidEntity.SPACE;
        double yPosInCell = dyPositive ? entity.getDeltaY() % FluidEntity.SPACE : FluidEntity.SPACE + entity.getDeltaY() % FluidEntity.SPACE;

        // Area of top right
        double bottomLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * (FluidEntity.SPACE - yPosInCell);

        // area of top left
        double bottomRightAreaInversion = xPosInCell * (FluidEntity.SPACE - yPosInCell);

        // area of bottom right
        double topLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * yPosInCell;

        // area of bottom left
        double topRightAreaInversion = xPosInCell * yPosInCell;

        double bottomLeftRatio = bottomLeftAreaInversion / FluidEntity.CELL_AREA;
        double bottomRightRatio = bottomRightAreaInversion / FluidEntity.CELL_AREA;
        double topLeftRatio = topLeftAreaInversion / FluidEntity.CELL_AREA;
        double topRightRatio = topRightAreaInversion / FluidEntity.CELL_AREA;

        transferTo(entity, t1x, t1y, bottomLeftRatio);
        transferTo(entity, t2x, t1y, bottomRightRatio);
        transferTo(entity, t1x, t2y, topLeftRatio);
        transferTo(entity, t2x, t2y, topRightRatio);
    }

    /**
     * https://en.wikipedia.org/wiki/Bilinear_interpolation
     */
    private void reverseAdvectionCellTransfer(int xIndex, int yIndex) {
        FluidEntity entity = entities[xIndex][yIndex];
        double negativeDeltaX = -entity.getDeltaX();
        double negativeDeltaY = -entity.getDeltaY();

        if (negativeDeltaX == 0 && negativeDeltaY == 0) return;

        int xIndexOffset = (int) negativeDeltaX / FluidEntity.SPACE;
        int yIndexOffset = (int) negativeDeltaY / FluidEntity.SPACE;

        boolean negativeDxPositive = negativeDeltaX > 0;
        boolean negativeDyPositive = negativeDeltaY > 0;

        int t1x = getLesserTargetIndex(xIndex, xIndexOffset, negativeDxPositive);
        int t1y = getLesserTargetIndex(yIndex, yIndexOffset, negativeDyPositive);

        int t2x = t1x + 1;
        int t2y = t1y + 1;

        double xPosInCell = negativeDxPositive ? -entity.getDeltaX() % FluidEntity.SPACE : FluidEntity.SPACE + (-entity.getDeltaX() % FluidEntity.SPACE);
        double yPosInCell = negativeDyPositive ? -entity.getDeltaY() % FluidEntity.SPACE : FluidEntity.SPACE + (-entity.getDeltaY() % FluidEntity.SPACE);


        // Area of top right
        double bottomLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * (FluidEntity.SPACE - yPosInCell);

        // area of top left
        double bottomRightAreaInversion = xPosInCell * (FluidEntity.SPACE - yPosInCell);

        // area of bottom right
        double topLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * yPosInCell;

        // area of bottom left
        double topRightAreaInversion = xPosInCell * yPosInCell;

        double bottomLeftRatio = bottomLeftAreaInversion / FluidEntity.CELL_AREA;
        double bottomRightRatio = bottomRightAreaInversion / FluidEntity.CELL_AREA;
        double topLeftRatio = topLeftAreaInversion / FluidEntity.CELL_AREA;
        double topRightRatio = topRightAreaInversion / FluidEntity.CELL_AREA;

        if (bottomLeftRatio > 1 || bottomRightRatio > 1 || topLeftRatio > 1 || topRightRatio > 1) {
            System.out.println("Math problem");
        }

        if ((bottomLeftRatio + bottomRightRatio + topLeftRatio + topRightRatio) - 1 > .001) {
            System.out.println("Math problem!");
        }

        transferFrom(entity, t1x, t1y, bottomLeftRatio);
        transferFrom(entity, t2x, t1y, bottomRightRatio);
        transferFrom(entity, t1x, t2y, topLeftRatio);
        transferFrom(entity, t2x, t2y, topRightRatio);
    }

    private void transferTo(FluidEntity originEntity, int targetXIndex, int targetYIndex, double ratio) {
        FluidEntity targetEntity = getEntity(targetXIndex, targetYIndex);
        originEntity.recordMassTransferTo(targetEntity, ratio);
    }

    private void transferFrom(FluidEntity targetEntity, int originXIndex, int originYIndex, double ratio) {
        FluidEntity originEntity = getEntity(originXIndex, originYIndex);
        originEntity.recordMassTransferTo(targetEntity, ratio);
    }

    private FluidEntity getEntity(int xIndex, int yIndex) {
        // Handle if outside of universe
        if (xIndex < 0) {
            xIndex = 0;
        } else if (xIndex >= entities.length) {
            xIndex = entities.length - 1;
        }
        if (yIndex < 0) {
            yIndex = 0;
        } else if (yIndex >= entities[xIndex].length) {
            yIndex = entities[xIndex].length - 1;
        }
        return entities[xIndex][yIndex];
    }

    private int getLesserTargetIndex(int sourceIndex, int indexOffset, boolean directionPositive) {
        return sourceIndex + indexOffset + (directionPositive ? 0 : -1);
    }

}

File: src/main/java/fluid/physics/Universe.kt
package fluid.physics

import fluid.entity.FluidEntity
import java.util.stream.IntStream

/**
 * http://cowboyprogramming.com/2008/04/01/practical-fluid-mechanics/
 * http://www.dgp.toronto.edu/people/stam/reality/Research/pdf/GDC03.pdf
 */
class Universe(val entities: Array<Array<FluidEntity>>) {
    private var step = 0

    /**
     * Run round of physics
     */
    @Synchronized
    fun updateUniverseState() {
        ExternalInput.applyInput(entities, step)
        incrementFluid()
        step++
    }

    private fun incrementFluid() {
        // force applications
        applySoloEffects()
        applyNeighborInteractions()
        IntStream.range(0, entities.size).forEach { x: Int -> IntStream.range(0, entities[x].size).forEach { y: Int -> entities[x][y].changeHeat() } }
        IntStream.range(0, entities.size).forEach { x: Int -> IntStream.range(0, entities[x].size).forEach { y: Int -> entities[x][y].changeForce() } }
        advection()

        // transfer application
        IntStream.range(0, entities.size).forEach { x: Int -> IntStream.range(0, entities[x].size).forEach { y: Int -> entities[x][y].convertMassTransferToAbsoluteChange() } }
        IntStream.range(0, entities.size).parallel().forEach { x: Int -> IntStream.range(0, entities[x].size).forEach { y: Int -> entities[x][y].changeMass() } }
    }

    fun applySoloEffects() {
        IntStream.range(0, entities.size).parallel().forEach { i: Int -> IntStream.range(0, entities[i].size).forEach { j: Int -> entities[i][j].applySoloEffects() } }
    }

    fun applyNeighborInteractions() {

        // Left entity
        IntStream.range(1, entities.size).parallel().forEach { i: Int ->
            IntStream.range(0, entities[i].size).forEach { j: Int ->
                val entity = entities[i][j]
                entity.applyNeighborInteractions(entities[i - 1][j])
            }
        }

        // Right entity
        IntStream.range(0, entities.size - 1).parallel().forEach { i: Int ->
            IntStream.range(0, entities[i].size).forEach { j: Int ->
                val entity = entities[i][j]
                entity.applyNeighborInteractions(entities[i + 1][j])
            }
        }

        // Lower entity
        IntStream.range(0, entities.size).parallel().forEach { i: Int ->
            IntStream.range(1, entities[i].size).forEach { j: Int ->
                val entity = entities[i][j]
                entity.applyNeighborInteractions(entities[i][j - 1])
            }
        }

        // Upper entity
        IntStream.range(0, entities.size).parallel().forEach { i: Int ->
            IntStream.range(0, entities[i].size - 1).forEach { j: Int ->
                val entity = entities[i][j]
                entity.applyNeighborInteractions(entities[i][j + 1])
            }
        }
    }

    /**
     * Advection moves the quantities from point to its connections/neighbors. Quantities include velocity/mass/heat/etc.
     * The amount moved from one point to another is based on the given point's velocity.
     */
    private fun advection() {
        IntStream.range(0, entities.size).parallel().forEach { i: Int ->
            IntStream.range(0, entities[i].size).forEach { j: Int ->
                forwardAdvectionCellTransfer(i, j)
                reverseAdvectionCellTransfer(i, j)
            }
        }
    }

    /**
     * https://en.wikipedia.org/wiki/Bilinear_interpolation
     */
    private fun forwardAdvectionCellTransfer(xIndex: Int, yIndex: Int) {
        val entity = entities[xIndex][yIndex]
        val deltaX = entity.deltaX
        val deltaY = entity.deltaY
        if (deltaX == 0.0 && deltaY == 0.0) return
        val xIndexOffset = deltaX.toInt() / FluidEntity.SPACE
        val yIndexOffset = deltaY.toInt() / FluidEntity.SPACE
        val dxPositive = deltaX > 0
        val dyPositive = deltaY > 0
        val t1x = getLesserTargetIndex(xIndex, xIndexOffset, dxPositive)
        val t1y = getLesserTargetIndex(yIndex, yIndexOffset, dyPositive)
        val t2x = t1x + 1
        val t2y = t1y + 1
        val xPosInCell = if (dxPositive) entity.deltaX % FluidEntity.SPACE else FluidEntity.SPACE + entity.deltaX % FluidEntity.SPACE
        val yPosInCell = if (dyPositive) entity.deltaY % FluidEntity.SPACE else FluidEntity.SPACE + entity.deltaY % FluidEntity.SPACE

        // Area of top right
        val bottomLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * (FluidEntity.SPACE - yPosInCell)

        // area of top left
        val bottomRightAreaInversion = xPosInCell * (FluidEntity.SPACE - yPosInCell)

        // area of bottom right
        val topLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * yPosInCell

        // area of bottom left
        val topRightAreaInversion = xPosInCell * yPosInCell
        val bottomLeftRatio = bottomLeftAreaInversion / FluidEntity.CELL_AREA
        val bottomRightRatio = bottomRightAreaInversion / FluidEntity.CELL_AREA
        val topLeftRatio = topLeftAreaInversion / FluidEntity.CELL_AREA
        val topRightRatio = topRightAreaInversion / FluidEntity.CELL_AREA
        transferTo(entity, t1x, t1y, bottomLeftRatio)
        transferTo(entity, t2x, t1y, bottomRightRatio)
        transferTo(entity, t1x, t2y, topLeftRatio)
        transferTo(entity, t2x, t2y, topRightRatio)
    }

    /**
     * https://en.wikipedia.org/wiki/Bilinear_interpolation
     */
    private fun reverseAdvectionCellTransfer(xIndex: Int, yIndex: Int) {
        val entity = entities[xIndex][yIndex]
        val negativeDeltaX = -entity.deltaX
        val negativeDeltaY = -entity.deltaY
        if (negativeDeltaX == 0.0 && negativeDeltaY == 0.0) return
        val xIndexOffset = negativeDeltaX.toInt() / FluidEntity.SPACE
        val yIndexOffset = negativeDeltaY.toInt() / FluidEntity.SPACE
        val negativeDxPositive = negativeDeltaX > 0
        val negativeDyPositive = negativeDeltaY > 0
        val t1x = getLesserTargetIndex(xIndex, xIndexOffset, negativeDxPositive)
        val t1y = getLesserTargetIndex(yIndex, yIndexOffset, negativeDyPositive)
        val t2x = t1x + 1
        val t2y = t1y + 1
        val xPosInCell = if (negativeDxPositive) -entity.deltaX % FluidEntity.SPACE else FluidEntity.SPACE + -entity.deltaX % FluidEntity.SPACE
        val yPosInCell = if (negativeDyPositive) -entity.deltaY % FluidEntity.SPACE else FluidEntity.SPACE + -entity.deltaY % FluidEntity.SPACE


        // Area of top right
        val bottomLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * (FluidEntity.SPACE - yPosInCell)

        // area of top left
        val bottomRightAreaInversion = xPosInCell * (FluidEntity.SPACE - yPosInCell)

        // area of bottom right
        val topLeftAreaInversion = (FluidEntity.SPACE - xPosInCell) * yPosInCell

        // area of bottom left
        val topRightAreaInversion = xPosInCell * yPosInCell
        val bottomLeftRatio = bottomLeftAreaInversion / FluidEntity.CELL_AREA
        val bottomRightRatio = bottomRightAreaInversion / FluidEntity.CELL_AREA
        val topLeftRatio = topLeftAreaInversion / FluidEntity.CELL_AREA
        val topRightRatio = topRightAreaInversion / FluidEntity.CELL_AREA
        if (bottomLeftRatio > 1 || bottomRightRatio > 1 || topLeftRatio > 1 || topRightRatio > 1) {
            println("Math problem")
        }
        if (bottomLeftRatio + bottomRightRatio + topLeftRatio + topRightRatio - 1 > .001) {
            println("Math problem!")
        }
        transferFrom(entity, t1x, t1y, bottomLeftRatio)
        transferFrom(entity, t2x, t1y, bottomRightRatio)
        transferFrom(entity, t1x, t2y, topLeftRatio)
        transferFrom(entity, t2x, t2y, topRightRatio)
    }

    private fun transferTo(originEntity: FluidEntity, targetXIndex: Int, targetYIndex: Int, ratio: Double) {
        val targetEntity = getEntity(targetXIndex, targetYIndex)
        originEntity.recordMassTransferTo(targetEntity, ratio)
    }

    private fun transferFrom(targetEntity: FluidEntity, originXIndex: Int, originYIndex: Int, ratio: Double) {
        val originEntity = getEntity(originXIndex, originYIndex)
        originEntity.recordMassTransferTo(targetEntity, ratio)
    }

    private fun getEntity(xIndex: Int, yIndex: Int): FluidEntity {
        // Handle if outside of universe
        var x = xIndex
        var y = yIndex
        if (xIndex < 0) {
            x = 0
        } else if (xIndex >= entities.size) {
            x = entities.size - 1
        }
        if (y < 0) {
            y = 0
        } else if (y >= entities[x].size) {
            y = entities[x].size - 1
        }
        return entities[x][y]
    }

    private fun getLesserTargetIndex(sourceIndex: Int, indexOffset: Int, directionPositive: Boolean): Int {
        return sourceIndex + indexOffset + if (directionPositive) 0 else -1
    }

    companion object {
        /**
         * Maximum number of immediate neighbors that each [FluidEntity] may have bidirectional interactions with.
         * Used to ensure that an entity does not transfer more than 100% of its heat/pressure/etc to its neighbors
         */
        const val MAX_NEIGHBORS = 4
        const val GRAVITATIONAL_CONSTANT = .0001
    }

}